dj/sql/parsing/ast.py

```diff
@@ -0,0 +1,517 @@
+
+
+class Expression(Node):
+    """an expression type simply for type checking"""
+
+
+@dataclass(eq=False)  # type: ignore
+class Named(Expression):
+    """An Expression that has a name"""
+
+    name: str
+    quote_style: Optional[str]
+
+    @property
+    def quoted_name(self) -> str:
+        """
+        get the name of the Named Node including the quotes if any
+        """
+        return f'{self.quote_style if self.quote_style else ""}{self.name}{self.quote_style if self.quote_style else ""}'  # pylint: disable=C0301
+
+    def alias_or_name(self) -> str:
+        """
+        get the name or alias of the node
+        """
+        if len(self.parents) == 1:
+            parent = tuple(self.parents)[0]
+            if isinstance(parent, Alias):
+                return parent.name
+        return self.name
+
+
+class Operation(Expression):
+    """a type to overarch types that operate on other expressions"""
+
+
+# pylint: disable=C0103
+class UnaryOpKind(Enum):
+    """the accepted unary operations"""
+
+    Plus = "+"
+    Minus = "-"
+    Not = "NOT"
+
+
+# pylint: enable=C0103
+
+
+@dataclass(eq=False)
+class UnaryOp(Operation):
+    """an operation that operates on a single expression"""
+
+    op: UnaryOpKind  # pylint: disable=C0103
+    expr: Expression
+
+    def __hash__(self) -> int:
+        return hash((UnaryOp, self.op))
+
+
+# pylint: disable=C0103
+class BinaryOpKind(Enum):
+    """the accepted binary operations"""
+
+    And = "AND"
+    Or = "OR"
+    Eq = "="
+    NotEq = "<>"
+    Gt = ">"
+    Lt = "<"
+    GtEq = ">="
+    LtEq = "<="
+    BitwiseOr = "|"
+    BitwiseAnd = "&"
+    BitwiseXor = "^"
+    Multiply = "*"
+    Divide = "/"
+    Plus = "+"
+    Minus = "-"
+    Modulo = "%"
+
+
+# pylint: enable=C0103
+
+
+@dataclass(eq=False)
+class BinaryOp(Operation):
+    """represents an operation that operates on two expressions"""
+
+    left: Expression
+    op: BinaryOpKind  # pylint: disable=C0103
+    right: Expression
+
+    def __hash__(self) -> int:
+        return hash((BinaryOp, self.op))
+
+
+@dataclass(eq=False)
+class Between(Operation):
+    """
+    a between statement
+    """
+
+    expr: Expression
+    low: Expression
+    high: Expression
+
+    def __hash__(self) -> int:
+        return hash((Between, self.low, self.high))
+
+
+@dataclass(eq=False)
+class Case(Expression):
+    """a case statement of branches"""
+
+    conditions: List[Expression]
+    else_result: Optional[Expression]
+    operand: Optional[Expression]
+    results: List[Expression]
+
+    def __hash__(self) -> int:
+        return id(self)
+
+
+@dataclass(eq=False)
+class Function(Named, Operation):
+    """represents a function used in a statement"""
+
+    args: List[Expression]
+
+    def __hash__(self) -> int:
+        return hash(Function)
+
+
+@dataclass(eq=False)  # type: ignore
+class Value(Expression):
+    """base class for all values number, string, boolean"""
+
+    value: Union[str, bool, float, int]
```

Comment:
Can this be `None`?

Reply:
Oh yes, we need to represent a NULL here. Using None may be an ok shortcut, otherwise we'd need a new type class.

Reply:
OK added the a node `IsNull` as an operator